src/Build/Xml/XmlReaderExtension.cs

```diff
@@ -71,15 +73,44 @@ public void Dispose()
             _stream?.Dispose();
         }
 
+        private static volatile PropertyInfo _normalizationPropertyInfo;
+
+        private static PropertyInfo GetNormalizationPropertyInfo(Type xmlReaderType)
+        {
+            BindingFlags bindingFlags = BindingFlags.NonPublic | BindingFlags.SetProperty | BindingFlags.Instance;
+            if (_normalizationPropertyInfo == null)
+            {
+                _normalizationPropertyInfo = xmlReaderType.GetProperty("Normalization", bindingFlags);
+            }
+            else if (xmlReaderType != _normalizationPropertyInfo.ReflectedType)
+            {
+                Debug.Fail("GetNormalizationPropertyInfo can only take one type");
+                return xmlReaderType.GetProperty("Normalization", bindingFlags);
+            }
+            return _normalizationPropertyInfo;
+        }
+
         private static XmlReader GetXmlReader(string file, StreamReader input, bool loadAsReadOnly, out Encoding encoding)
         {
             string uri = new UriBuilder(Uri.UriSchemeFile, string.Empty) { Path = file }.ToString();
 
-            
-            // Ignore loadAsReadOnly for now; using XmlReader.Create results in whitespace changes
-            // of attribute text, specifically newline removal.
-            // https://github.com/Microsoft/msbuild/issues/4210
-            XmlReader reader = new XmlTextReader(uri, input) { DtdProcessing = DtdProcessing.Ignore };
+            XmlReader reader;
+            if (loadAsReadOnly)
+            {
+                XmlReaderSettings xrs = new XmlReaderSettings
+                {
+                    DtdProcessing = DtdProcessing.Ignore,
+                    IgnoreComments = true,
+                    IgnoreWhitespace = true,
+                };
+                reader = XmlReader.Create(input, xrs, uri);
+                // HACK: Set Normalization to false to behave the same as XmlTextReader.
+                GetNormalizationPropertyInfo(reader.GetType()).SetValue(reader, false);
```

Comment:
I would just create a local for PropertyInfo returned by `GetNormalizationPropertyInfo`, and check that for null, and fallback to the old behavior if it is null, otherwise call SetValue. Don't think there's a need for try/catch. Maybe also cache the result of `reader.GetType()`?

Doubt it's worth creating a delegate here (reflection is slow, but we're only calling it once per file, so doubt it will show in profiles at all). Maybe worth a try if we're being completely thorough.

Reply:
Done, thank you!